blocks/theano_expressions.py

```diff
@@ -14,11 +14,11 @@ def l2_norm(tensors):
         The tensors.
 
     """
-    flattened = [tensor.as_tensor_variable(t).flatten() for t in tensors]
-    flattened = [(t if t.ndim > 0 else t.dimshuffle('x'))
-                 for t in flattened]
-    joined = tensor.join(0, *flattened)
-    return tensor.sqrt(tensor.sqr(joined).sum())
+    grand_sum = 0.0
```

Comment:
Hi,

1. I need to fix a bug with the ifelse -> tensor.switch change 
(something about broadcasting).

2. What you propose is also OK, the main problem here is poor CAReduce 
performance. This improves things a little bit, but the real fix is 
here: 
https://github.com/janchorowski/Theano/commit/2e85535b36c90547bfa89bf954db8223dc935ed2

Which gets this 20 time faster.

Dmitry Bogdanov wrote:
>
> In blocks/theano_expressions.py
> <https://github.com/mila-udem/blocks/pull/781#discussion_r36042933>:
>
> > @@ -14,11 +14,11 @@ def l2_norm(tensors):
> > The tensors.
> >
> > """
> > - flattened = [tensor.as_tensor_variable(t).flatten() for t in tensors]
> > - flattened = [(t if t.ndim> 0 else t.dimshuffle('x'))
> > - for t in flattened]
> > - joined = tensor.join(0, *flattened)
> > - return tensor.sqrt(tensor.sqr(joined).sum())
> > + grand_sum = 0.0
>
> We had quite a discussion once with @bartvm
> <https://github.com/bartvm> about the right way to sum here. The
> method you propose has a disadvantage that the computation graph will
> be as deep as many tensor are given as arguments. This might hurt at
> compile time and also makes it harder to investigate debugprints.
>
> On the other hand, I agree that the old implementation is memory
> inefficient.
>
> It seems like one can combine the advantages of two by creating a
> vector of scalars in the way you do it and joining these scalars like
> we did. What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/781/files#r36042933>.